### PR TITLE
impl(spanner): start second emulator to service REST transport

### PR DESCRIPTION
**Pull request by scotthart**
Part of the work to test gRPC/HTTP transcoding using generated Spanner admin code.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10704)

### CODE REVIEW

File: google/cloud/spanner/ci/lib/spanner_emulator.sh [hunk 1]
```diff
@@ -25,6 +25,7 @@ source module /ci/lib/io.sh
 # Global variable that holds the PID of the Spanner emulator. This will be set
 # when the emulator is started, and it will be used to kill the emulator.
```
Comment (devbww) [inline google/cloud/spanner/ci/lib/spanner_emulator.sh:26]: Pluralize?

File: google/cloud/spanner/ci/lib/spanner_emulator.sh [hunk 2]
```diff
@@ -83,6 +102,36 @@ function spanner_emulator::start() {
...
+  local http_port=$((emulator_port + 1))
+  local grpc_port=$((http_port + 1))
+  "${SPANNER_HTTP_EMULATOR_CMD}" --hostname "localhost" --grpc_port "${grpc_port}" --http_port "${http_port}" >http_emulator.log 2>&1 </dev/null &
+  SPANNER_HTTP_EMULATOR_PID=$!
+
+  http_emulator_port="$(spanner_emulator::internal::read_http_emulator_port http_emulator.log)"
+  if [[ "${http_emulator_port}" = "0" ]]; then
+    io::log_red "Cannot determine Cloud Spanner HTTP emulator port." >&2
+    cat http_emulator.log >&2
+    spanner_http_emulator::kill
```
Comment (devbww) [inline google/cloud/spanner/ci/lib/spanner_emulator.sh:128]: Does this exist?

File: google/cloud/spanner/ci/lib/spanner_emulator.sh [hunk 3]
```diff
@@ -83,6 +102,36 @@ function spanner_emulator::start() {
 
   export SPANNER_EMULATOR_HOST="localhost:${emulator_port}"
   io::log "Spanner emulator started at ${SPANNER_EMULATOR_HOST}"
+
+  # Repeat the process to launch the emulator with HTTP support. We launch a separate process
+  # as existing tests fail if we try and use gateway_main for both gRPC and REST.
+  readonly SPANNER_HTTP_EMULATOR_CMD="${CLOUD_SDK_LOCATION}/bin/cloud_spanner_emulator/gateway_main"
+  echo "${SPANNER_HTTP_EMULATOR_CMD}"
```
Comment (devbww) [inline google/cloud/spanner/ci/lib/spanner_emulator.sh:109]: Debugging?

File: google/cloud/spanner/ci/lib/spanner_emulator.sh [hunk 4]
```diff
@@ -83,6 +102,36 @@ function spanner_emulator::start() {
...
+
+  http_emulator_port="$(spanner_emulator::internal::read_http_emulator_port http_emulator.log)"
+  if [[ "${http_emulator_port}" = "0" ]]; then
+    io::log_red "Cannot determine Cloud Spanner HTTP emulator port." >&2
+    cat http_emulator.log >&2
+    spanner_http_emulator::kill
+    return 1
+  fi
+
+  # Attempting to use https:// results in SSL errors.
```
Comment (devbww) [inline google/cloud/spanner/ci/lib/spanner_emulator.sh:132]: s/Attempting to use/Using/?

File: google/cloud/spanner/ci/lib/spanner_emulator.sh [hunk 5]
```diff
@@ -83,6 +102,36 @@ function spanner_emulator::start() {
...
+  # The tests typically run in a Docker container, where the ports are largely
+  # free; when using in manual tests, you can set REST_EMULATOR_PORT.
+  rm -f http_emulator.log
+  # Leverage the previously determined emulator_port to avoid collision.
+  local http_port=$((emulator_port + 1))
+  local grpc_port=$((http_port + 1))
+  "${SPANNER_HTTP_EMULATOR_CMD}" --hostname "localhost" --grpc_port "${grpc_port}" --http_port "${http_port}" >http_emulator.log 2>&1 </dev/null &
+  SPANNER_HTTP_EMULATOR_PID=$!
+
+  http_emulator_port="$(spanner_emulator::internal::read_http_emulator_port http_emulator.log)"
```
Comment (devbww) [inline google/cloud/spanner/ci/lib/spanner_emulator.sh:124]: For `emulator_main` we re-used the local `emulator_port`.  Should we just do the same with `http_port` here?  Or rename `http_port` to `http_emulator_port` (and `grpc_port`)?

File: google/cloud/spanner/ci/lib/spanner_emulator.sh [hunk 6]
```diff
@@ -83,6 +102,36 @@ function spanner_emulator::start() {
...
+    return 1
+  fi
+
+  # The tests typically run in a Docker container, where the ports are largely
+  # free; when using in manual tests, you can set REST_EMULATOR_PORT.
+  rm -f http_emulator.log
+  # Leverage the previously determined emulator_port to avoid collision.
+  local http_port=$((emulator_port + 1))
+  local grpc_port=$((http_port + 1))
+  "${SPANNER_HTTP_EMULATOR_CMD}" --hostname "localhost" --grpc_port "${grpc_port}" --http_port "${http_port}" >http_emulator.log 2>&1 </dev/null &
```
Comment (devbww) [inline google/cloud/spanner/ci/lib/spanner_emulator.sh:121]: Line wrap?

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10704)* status: 0 of 2 files reviewed, 6 unresolved discussions (waiting on @devbww)

___
*[`google/cloud/spanner/ci/lib/spanner_emulator.sh` line 26 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10704#-NN87b681aJ89d9VV1Yn-r1-26:-NN8Lpm08alx-p9ZVZAp:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/bcf221eafa0db1decf00be9418eda5c83f1271e5/google/cloud/spanner/ci/lib/spanner_emulator.sh#L26)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Pluralize?
</blockquote></details>

Done.
___
*[`google/cloud/spanner/ci/lib/spanner_emulator.sh` line 109 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10704#-NN87b681aJ89d9VV1Yn-r1-109:-NN8Lq8I25g2gxk3otey:bss4a2b) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/bcf221eafa0db1decf00be9418eda5c83f1271e5/google/cloud/spanner/ci/lib/spanner_emulator.sh#L109)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Debugging?
</blockquote></details>

Removed.
___
*[`google/cloud/spanner/ci/lib/spanner_emulator.sh` line 121 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10704#-NN87b681aJ89d9VV1Yn-r1-121:-NN8LrtC0yPI110xBjOA:bqmjwcz) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/bcf221eafa0db1decf00be9418eda5c83f1271e5/google/cloud/spanner/ci/lib/spanner_emulator.sh#L121)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Line wrap?
</blockquote></details>

Added.
___
*[`google/cloud/spanner/ci/lib/spanner_emulator.sh` line 124 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10704#-NN87b681aJ89d9VV1Yn-r1-124:-NN8Lz8iALhk3BIXEMpx:b-28tnsf) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/bcf221eafa0db1decf00be9418eda5c83f1271e5/google/cloud/spanner/ci/lib/spanner_emulator.sh#L124)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

For `emulator_main` we re-used the local `emulator_port`.  Should we just do the same with `http_port` here?  Or rename `http_port` to `http_emulator_port` (and `grpc_port`)?
</blockquote></details>

I think I understood your suggestion...
___
*[`google/cloud/spanner/ci/lib/spanner_emulator.sh` line 128 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10704#-NN87b681aJ89d9VV1Yn-r1-128:-NN8LuWs95ghPsfx71o5:bxiphsy) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/bcf221eafa0db1decf00be9418eda5c83f1271e5/google/cloud/spanner/ci/lib/spanner_emulator.sh#L128)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Does this exist?
</blockquote></details>

Corrected to `spanner_emulator:kill`
___
*[`google/cloud/spanner/ci/lib/spanner_emulator.sh` line 132 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10704#-NN87b681aJ89d9VV1Yn-r1-132:-NN8Ly_w9KyW76ddbdxf:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/bcf221eafa0db1decf00be9418eda5c83f1271e5/google/cloud/spanner/ci/lib/spanner_emulator.sh#L132)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

s/Attempting to use/Using/?
</blockquote></details>

Done.

**Review by devbww — APPROVED** [review]
> Reviewed 1 of 2 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10704)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @scotthart)

___
*[`google/cloud/spanner/ci/lib/spanner_emulator.sh` line 26 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10704#-NN87b681aJ89d9VV1Yn-r1-26:-NN8N17e3HvUKP8rEnKe:bp0o3pt) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/bcf221eafa0db1decf00be9418eda5c83f1271e5/google/cloud/spanner/ci/lib/spanner_emulator.sh#L26)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

s/Global variable that holds/Global variables that hold/

**Pull request merged**